### ISSUE TITLE
Poor performance on macOS

### ISSUE THREAD

**Issue by mrbell321** [pyserial/pyserial#515]
I am attempting to communicate w/ a device that speaks ASCII over RS-232. You send commands and it echos back on success. There are several hundred kilobytes of commands that need to be issued.
I was seeing poor performance. I have tracked it, as best I can, do pyserial.

I have written the following test to illustrate the issue:
```
def pytest_generate_tests(metafunc):
    if "baudrate" in metafunc.fixturenames:
        metafunc.parametrize("baudrate", (baudrate for baudrate in Serial.BAUDRATES if 9600 <= baudrate <= 115200))
    if "testdata" in metafunc.fixturenames:
        bs_r = [(bs, 2**r) for r in range(3, 8, 2) for bs in (b"U", b"3", b"w")]
        metafunc.parametrize("testdata", (bs * r for bs, r in bs_r),
                                     ids=(f"{bs} * {r}" for bs, r in bs_r))

def test_loopback_direct(testdata, baudrate):
    serial = Serial(port='/dev/cu.usbserial-FTCAY8UZ')
    serial.baudrate = baudrate
    serial.flushOutput()
    serial.flushInput()

    def write(testdata: bytes) -> None:
        serial.write(testdata)
        serial.flushOutput()

    def read(n: int) -> bytes:
        start_r = time.monotonic()
        result = bytearray()
        while len(result) < n:
            result.extend(serial.read(serial.in_waiting))
        return bytes(result)

    def timed(func, *args):
        start_mt = time.monotonic()
        result = func(*args)
        end_mt = time.monotonic()

        return result, end_mt - start_mt

    _, write_time = timed(write, testdata)
    actual, read_time = timed(read, len(testdata))

    print(f"ch: '{chr(testdata[0])}'; sz: {len(testdata)}; baud: {baudrate}; tx: {write_time}; rx: {read_time}")
    assert actual == testdata
````

This will try to send and read 8, 32, and 128 byte strings at every baud rate between 9600 and 115200(other bauds supported by pyserial are not supported by my hardware or project). Strings are comprised of 'U' is 0b01010101, '3' is 0b00110011, and 'w' is 0b01110111, so basically different timings of alternating 1's and 0's.

It's expecting to read exactly what it writes because I've wired the RS-232 port(an FTDI USB-RS232 dongle) with a physical loopback(connect TX to RX). I used a physical loopback to remove any concern about turnaround time from a remote device.

Lockups(as in it gets stuck in the read forever, meaning it's missed some data that was physically there) tend to happen above 115200. Luckily for me, my application maxes out at 115200, so I'm not less concerned, but it's an interesting datapoint. I also get lockups below 115200, but much less frequently.

The more problematic issue is speed:
```
ch: '3'; sz: 8; baud: 9600; tx: 0.00021691100000009733; rx: 0.26131050899999986
ch: '3'; sz: 32; baud: 9600; tx: 0.00026775699999981306; rx: 0.28544173500000003
ch: '3'; sz: 128; baud: 9600; tx: 0.0003735080000000224; rx: 0.38507877300000004
ch: 'U'; sz: 8; baud: 9600; tx: 0.0003041390000000366; rx: 0.261603931
ch: 'U'; sz: 32; baud: 9600; tx: 0.00029090999999992206; rx: 0.28551700699999993
ch: 'U'; sz: 128; baud: 9600; tx: 0.00041519800000022755; rx: 0.38469731100000004
ch: 'w'; sz: 8; baud: 9600; tx: 0.00029173200000021104; rx: 0.260977236
ch: 'w'; sz: 32; baud: 9600; tx: 0.0002677659999998916; rx: 0.28654258899999974
ch: 'w'; sz: 128; baud: 9600; tx: 0.0003099739999998796; rx: 0.3854876680000001
ch: '3'; sz: 8; baud: 19200; tx: 0.000292234999999863; rx: 0.25674410999999964
ch: '3'; sz: 32; baud: 19200; tx: 0.000306768999999818; rx: 0.2694930219999998
ch: '3'; sz: 128; baud: 19200; tx: 0.00032810600000043877; rx: 0.319217632
ch: 'U'; sz: 8; baud: 19200; tx: 0.00034149099999991606; rx: 0.25650378399999996
ch: 'U'; sz: 32; baud: 19200; tx: 0.0003206829999999883; rx: 0.2691566559999998
ch: 'U'; sz: 128; baud: 19200; tx: 0.00034597099999977843; rx: 0.3192357640000001
ch: 'w'; sz: 8; baud: 19200; tx: 0.0002986969999998479; rx: 0.25695754000000015
ch: 'w'; sz: 32; baud: 19200; tx: 0.00032339799999991925; rx: 0.26902464800000025
ch: 'w'; sz: 128; baud: 19200; tx: 0.00033933599999969033; rx: 0.3189014330000006
ch: '3'; sz: 8; baud: 38400; tx: 0.00022700199999992066; rx: 0.25499005100000005
ch: '3'; sz: 32; baud: 38400; tx: 0.0002943580000005497; rx: 0.2615028230000007
ch: '3'; sz: 128; baud: 38400; tx: 0.0003183870000000866; rx: 0.28616120000000045
ch: 'U'; sz: 8; baud: 38400; tx: 0.0003507960000002086; rx: 0.2546732949999999
ch: 'U'; sz: 32; baud: 38400; tx: 0.0003138900000001499; rx: 0.2611683600000001
ch: 'U'; sz: 128; baud: 38400; tx: 0.0003122150000001156; rx: 0.2860405180000001
ch: 'w'; sz: 8; baud: 38400; tx: 0.0002870059999997565; rx: 0.25501268200000027
ch: 'w'; sz: 32; baud: 38400; tx: 0.00030984999999983387; rx: 0.26141939100000045
ch: 'w'; sz: 128; baud: 38400; tx: 0.000311405999999792; rx: 0.2860767909999993
ch: '3'; sz: 8; baud: 57600; tx: 0.00030658100000025; rx: 0.25391349999999946
ch: '3'; sz: 32; baud: 57600; tx: 0.0002960700000009808; rx: 0.2584686759999997
ch: '3'; sz: 128; baud: 57600; tx: 0.00032355299999942133; rx: 0.27527845299999854
ch: 'U'; sz: 8; baud: 57600; tx: 0.00028539300000041123; rx: 0.25441498100000004
ch: 'U'; sz: 32; baud: 57600; tx: 0.0003074089999994811; rx: 0.2584645529999996
ch: 'U'; sz: 128; baud: 57600; tx: 0.0003821149999989615; rx: 0.2750927839999999
ch: 'w'; sz: 8; baud: 57600; tx: 0.000298542000001234; rx: 0.25366186699999993
ch: 'w'; sz: 32; baud: 57600; tx: 0.0002962950000000575; rx: 0.25837463300000074
ch: 'w'; sz: 128; baud: 57600; tx: 0.0003059309999997595; rx: 0.2749769889999989
ch: '3'; sz: 8; baud: 115200; tx: 0.0002932970000000257; rx: 0.2534307789999999
ch: '3'; sz: 32; baud: 115200; tx: 0.0002898289999999637; rx: 0.256054969
ch: '3'; sz: 128; baud: 115200; tx: 0.00030081699999984224; rx: 0.26350622099999965
ch: 'U'; sz: 8; baud: 115200; tx: 0.000226523000000034; rx: 0.25339219700000004
ch: 'U'; sz: 32; baud: 115200; tx: 0.00024340499999997434; rx: 0.2553583530000001
ch: 'U'; sz: 128; baud: 115200; tx: 0.0003089139999996604; rx: 0.263742664
ch: 'w'; sz: 8; baud: 115200; tx: 0.000281980000000015; rx: 0.2536936110000001
ch: 'w'; sz: 32; baud: 115200; tx: 0.0003012570000002768; rx: 0.25522577600000007
ch: 'w'; sz: 128; baud: 115200; tx: 0.0003105729999997031; rx: 0.26431345100000003
```
Notice that the read times are nearly the same whether you are reading 8bytes at 9600 baud or 128bytes at 115200 baud. 0.25 is a long time to wait for every response when sending several hundred k of ~8-byte commands

Write times also have a lower limit, but it's FAR lower and is not my current bottle neck. 

I can use the same same dongle w/ other(non-python) tools and I do not see this delay.

`Python 3.7.3 (default, Mar 27 2019, 09:23:15) 
[Clang 10.0.1 (clang-1001.0.46.3)] on darwin`
and
`pyserial                                3.4`

**Comment by mrbell321** [comment]
Interestingly, I added some time tracking to miniterm and read times are closer to 0.001s @9600 and drop as baud rate increases. @115200, complete read-write cycles for an immediate hardware echo are are seem to average about 0.0005s. This is far more acceptable. 

I suppose this could be due to the threading used in miniterm and starting to read before sending writing as compared to the single threaded write then read nature of my sample test. Is there significant overhead in simply calling read() that is sidestepped by calling it during initialization?

I actually started this investigation on pyserial-asyncio and was having the same slowness, so I backed up to regular pyserial to see if it was inherent in python/pyserial or the asyncio framework. Since I'm seeing single threaded slowness in both asyncio and synchronous code but not in threaded, it seems like the only solution is to use threads?

**Comment by mrbell321** [comment]
OK, well I went back and re-ran my original tests and it's working fine. I spent all week last week trying to track that down and it fixed itself over the weekend. 
Closing this until I run into it again.

**Comment by mrbell321** [comment]
For posterity:
I experienced this again after the last Mac OS update. I believe this is an issue w/ Apple's FTDI driver. Installing the VCP driver from FTDI resolved it.
Prior to that, however, I did a little more experimentation and the issue w/ Apple's driver seems to be in reading that last character from the buffer. If, for example, there are 10 characters in the input buffer, reading to end of line, reading 10 characters at once takes at least 0.25s regardless of baud(or even number of characters, to a very large point). Reading the first 9 characters(singly or as a group) takes hardly any time at all, as you would expect, but that last character takes about 0.25s.

BTW, installing FTDI's driver gives a warning that the driver will be incompatible w/ future versions which will probably deter some people.
More importantly, this must be taken into consideration for distribution.
